### ISSUE TITLE
Aftermath SDK linker issues when compiling micromesh_tool

### ISSUE THREAD

**Issue by pknowlesnv** [NVIDIAGameWorks/Displacement-MicroMap-Toolkit#6]
Splitting out the Aftermath SDK linker issue reported [here](https://github.com/NVIDIAGameWorks/Displacement-MicroMap-Toolkit/issues/5#issuecomment-1570586349) to this issue.

```
nsight_aftermath_vk.obj : error LNK2019: unresolved external symbol GFSDK_Aftermath_EnableGpuCrashDumps referenced in function "public: __cdecl <lambda_ada5e3fd341686269c3305092ae58032>::operator()(void)const " (??R<lambda_ada5e3fd341686269c3305092ae58032>@@QEBA@XZ)
nsight_aftermath_vk.obj : error LNK2019: unresolved external symbol GFSDK_Aftermath_DisableGpuCrashDumps referenced in function "public: __cdecl nvvk::GpuCrashTrackerImpl::~GpuCrashTrackerImpl(void)" (??1GpuCrashTrackerImpl@nvvk@@QEAA@XZ)
nsight_aftermath_vk.obj : error LNK2019: unresolved external symbol GFSDK_Aftermath_GpuCrashDump_CreateDecoder referenced in function "public: __cdecl <lambda_bcee80718c1a11da9ec69d89ad3ff146>::operator()(void)const " (??R<lambda_bcee80718c1a11da9ec69d89ad3ff146>@@QEBA@XZ)
nsight_aftermath_vk.obj : error LNK2019: unresolved external symbol GFSDK_Aftermath_GpuCrashDump_DestroyDecoder referenced in function "public: __cdecl <lambda_5faf1e2e40cca68f5896d5afe79305e1>::operator()(void)const " (??R<lambda_5faf1e2e40cca68f5896d5afe79305e1>@@QEBA@XZ)
nsight_aftermath_vk.obj : error LNK2019: unresolved external symbol GFSDK_Aftermath_GpuCrashDump_GetBaseInfo referenced in function "public: __cdecl <lambda_f9bd0c1c9d47cff75d36e711b52c0aaf>::operator()(void)const " (??R<lambda_f9bd0c1c9d47cff75d36e711b52c0aaf>@@QEBA@XZ)
nsight_aftermath_vk.obj : error LNK2019: unresolved external symbol GFSDK_Aftermath_GpuCrashDump_GetDescriptionSize referenced in function "public: __cdecl <lambda_3c95f9426cc46d89cb3190e35c79b5f5>::operator()(void)const " (??R<lambda_3c95f9426cc46d89cb3190e35c79b5f5>@@QEBA@XZ)
nsight_aftermath_vk.obj : error LNK2019: unresolved external symbol GFSDK_Aftermath_GpuCrashDump_GetDescription referenced in function "public: __cdecl <lambda_f67ede3340a473d3d5537a9f3d1bec42>::operator()(void)const " (??R<lambda_f67ede3340a473d3d5537a9f3d1bec42>@@QEBA@XZ)
nsight_aftermath_vk.obj : error LNK2019: unresolved external symbol GFSDK_Aftermath_GpuCrashDump_GenerateJSON referenced in function "public: __cdecl <lambda_1dcf888cf20408fc8fdac84a0503d065>::operator()(void)const " (??R<lambda_1dcf888cf20408fc8fdac84a0503d065>@@QEBA@XZ)
nsight_aftermath_vk.obj : error LNK2019: unresolved external symbol GFSDK_Aftermath_GpuCrashDump_GetJSON referenced in function "public: __cdecl <lambda_3cd638047a4cd1bc5991cda8dad8e072>::operator()(void)const " (??R<lambda_3cd638047a4cd1bc5991cda8dad8e072>@@QEBA@XZ)
nsight_aftermath_vk.obj : error LNK2019: unresolved external symbol GFSDK_Aftermath_GetShaderDebugInfoIdentifier referenced in function "public: __cdecl <lambda_d7c784221bde71420e25ac6016eac910>::operator()(void)const " (??R<lambda_d7c784221bde71420e25ac6016eac910>@@QEBA@XZ)
nsight_aftermath_vk.obj : error LNK2019: unresolved external symbol GFSDK_Aftermath_GetShaderHashSpirv referenced in function "public: __cdecl <lambda_ea381bc51ee967ea9bd4702e49697a1e>::operator()(void)const " (??R<lambda_ea381bc51ee967ea9bd4702e49697a1e>@@QEBA@XZ)
nsight_aftermath_vk.obj : error LNK2019: unresolved external symbol GFSDK_Aftermath_GetShaderDebugNameSpirv referenced in function "public: __cdecl <lambda_49f9c48f787e6ec561d4883cad00912c>::operator()(void)const " (??R<lambda_49f9c48f787e6ec561d4883cad00912c>@@QEBA@XZ)
D:\Personal\Projects\ProjMicromap\build\micromesh_tool\Debug\micromesh_tool.exe : fatal error LNK1120: 12 unresolved externals
```

**Comment by pknowlesnv** [comment]
I'm hoping this only happens when attempting to compile micromesh_tool with the aftermath SDK. That is, unless the `NSIGHT_AFTERMATH_SDK` path is set in the cmake config, aftermath related code should not be compiled.

Assuming aftermath is needed...

The `micromesh_toolkit` cmake files have their own method to build with aftermath - copying/symlinking the inlude and lib file here https://github.com/NVIDIAGameWorks/Displacement-MicroMap-Toolkit/blob/main/micromesh_toolbox/thirdparty/aftermath_sdk/_readme.txt. Unfortunately this is separate from the nvpro_core aftermath integration. The nvpro_core way is to define the path to the aftermath SDK:

```
cmake -DNSIGHT_AFTERMATH_SDK=/path/to/aftermath_sdk ...
```

The configure step should print something like "Looking for nSight Aftermath at: <path>". For Displacement-MicroMap-Toolkit, the top level CMakeLists.txt file may need the following uncommented too:

```
- # _add_package_NsightAftermath()
+ _add_package_NsightAftermath()
```

The unresolved symbols point to `GFSDK_Aftermath_Lib.x64.lib` not making it to the linker (micromesh_tool -> properties -> Linker -> Input -> Additional Dependencies), which should be set by cmake. It looks like nvpro_core expects `LIBRARIES_OPTIMIZED` to be added to projects that use it and both `micromesh_tool` and `dmm_displacement` are missing this.

```
- target_link_libraries(${PROJECT_NAME} micromesh_tool_lib)
+ target_link_libraries(${PROJECT_NAME} micromesh_tool_lib ${LIBRARIES_OPTIMIZED})
```

```
- target_link_libraries (${PROJECT_NAME} ${PLATFORM_LIBRARIES})
+ target_link_libraries (${PROJECT_NAME} ${PLATFORM_LIBRARIES} ${LIBRARIES_OPTIMIZED})
```

When launching, `GFSDK_Aftermath_Lib.x64.dll` and possibly other binaries from the aftermath SDK will be needed. When debugging in VS the working directory is `$(ProjectDir)`, e.g. where `micromesh_tool.vcxproj` is in cmake's build directory, and you could copy the dlls there.

<img src="https://github.com/NVIDIAGameWorks/Displacement-MicroMap-Toolkit/assets/22082237/85e2a12f-5cb1-4b56-8ce1-856a097d8b8d" width="400px" />

The app should now print "WARNING: Aftermath extensions enabled. This may affect performance." when launching. When a crash happens, aftermath provides a callback to write a `.nvdbg`. Sticking a breakpoint there to know it's called and the call finishes is handy - in this case, `GpuCrashTrackerImpl::writeShaderDebugInformationToFile()`.

For a test case, I tried the arguments `--output sphere_l5.gltf {generate --resolution 1024 --geometry sphere} {pretessellate} {bake --level 5}`. To make it crash I commented out the line `sceneDescBuf = m_vk.resAllocator->createBuffer(...)` in `meshops_bake_vk.cpp`. After aborting from a `VK_ERROR_DEVICE_LOST` there should be a `micromesh_tool-*.nv-gpudmp` in the project directory that can be opened in Nsight Graphics. In the Crash Info tab, clicking on the faulted warp should show the location in the Shader Source panel on the right (this is read from the .nvdbg file).

```
    BakerMeshInfos  meshes      = BakerMeshInfos(sceneDesc.baseMeshAddress);
->  BakerMeshInfo   mesh        = meshes.i[0];
```

Leaving this issue open for now to track making the `_add_package_NsightAftermath()` and `${LIBRARIES_OPTIMIZED}` changes or otherwise streamlining the aftermath process.

**Comment by pknowlesnv** [comment]
This should now be fixed in aaeeebd67891fe7d5bb60dfce3a582d31c80c7cf. Closing.
